### PR TITLE
Add debug logging for snapshots

### PR DESCRIPTION
**Pull request by DiannaHohensee**
"Specifically around pausing shard snapshots on
  node removal, and finalizing shards snapshots
  that change the shard generation of other non-
  finalized snapshots

 Relates ES-8566."

-------------------

Related to ES-8566. Trying to get better logging when/if that issue reoccurs.

Not certain about the log-level. I was thinking debug and then changing the serverless log level defaults? Not sure whether that's reasonable. Or maybe I just make them info-level: we shouldn't encounter the logging except when scaling up/down removes nodes; and then there will be logging when multiple snapshots of the same shard are running, which I would expect to be more rare.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java [hunk 1]
```diff
@@ -207,6 +211,12 @@ public Map<RepositoryShardId, Set<ShardGeneration>> obsoleteGenerations(String r
                     // We moved from a non-null generation successful generation to a different non-null successful generation
                     // so the original generation is clearly obsolete because it was in-flight before and is now unreferenced everywhere.
                     obsoleteGenerations.computeIfAbsent(repositoryShardId, ignored -> new HashSet<>()).add(oldStatus.generation());
+                    logger.debug(
+                        "Marking shard generation [{}] file for cleanup. The finalized shard generation is now [{}] for shard snapshot [{}]",
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:215]: I think we should include the shard ID here too

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java [hunk 2]
```diff
@@ -442,6 +452,9 @@ public SnapshotsInProgress withUpdatedNodeIdsForRemoval(ClusterState clusterStat
 
         // remove any nodes which are no longer marked for shutdown if they have no running shard snapshots
         updatedNodeIdsForRemoval.removeAll(getObsoleteNodeIdsForRemoval(nodeIdsMarkedForRemoval));
+        var restoredNodeIds = getObsoleteNodeIdsForRemoval(nodeIdsMarkedForRemoval);
+        updatedNodeIdsForRemoval.removeAll(restoredNodeIds);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:456]: No need to `removeAll()` twice here I think?

File: server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java [hunk 3]
```diff
@@ -109,6 +113,11 @@ public Map<IndexId, Map<Integer, ShardGeneration>> obsoleteShardGenerations(Shar
                 // Since this method assumes only additions and no removals of shards, a null updated generation means no update
                 if (updatedGeneration != null && oldGeneration != null && oldGeneration.equals(updatedGeneration) == false) {
                     obsoleteShardIndices.put(i, oldGeneration);
+                    logger.debug(
+                        "Marking snapshot generation [{}] for cleanup. The new generation is [{}]",
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:117]: Likewise here, we probably want to identify the shard in this message.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java [hunk 4]
```diff
@@ -605,9 +605,13 @@ private void syncShardStatsOnNewMaster(List<SnapshotsInProgress.Entry> entries)
                             );
                         } else if (stage == Stage.PAUSED) {
                             // but we think the shard has paused - we need to make new master know that
-                            logger.debug("""
-                                [{}] new master thinks the shard [{}] is still running but the shard paused locally, updating status on \
-                                master""", snapshot.snapshot(), shardId);
+                            logger.debug(
+                                "new master thinks that shard [{}] snapshot [{}], with shard generation [{}], is still running, but the "
+                                    + "shard snapshot is paused locally, updating status on master",
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:610]: nit, would prefer to keep the format string as a `"""multiline"""` one

**Review by DaveCTurner — COMMENTED** [review]
Seems ok to me, just a few small suggestions

**Comment by DaveCTurner** [comment]
> I would like to add logging when we restart/resume paused shard snapshots, as well, if you could point out where that happen -- it's unclear to me whether that's distinguishable from regularly starting a shard snapshot.

I think you're looking for `org.elasticsearch.snapshots.SnapshotsService#initShardSnapshotStatus`.

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java (hunk 1 above)
Comment (DiannaHohensee) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java]: I don't see a way to get the Shard ID here. Repository ID is simple. But it looks like the InProgressSnapshots.Entry has only a shards map of ShardID->ShardSnapshotStatus, which doesn't do much good here.

**Comment by DiannaHohensee** [comment]
>> Or maybe I just make them info-level:

> No I think this will be far too verbose for INFO, we're adding O(#shards) messages here.

The pausing part would be O(#shards) messages, but how about the finalization override of shard generation? Are concurrent shard snapshots that common?

**Comment by DiannaHohensee** [comment]
>> I would like to add logging when we restart/resume paused shard snapshots, as well, if you could point out where that happen -- it's unclear to me whether that's distinguishable from regularly starting a shard snapshot.

> I think you're looking for org.elasticsearch.snapshots.SnapshotsService#initShardSnapshotStatus.

This appears to set the current state for a shard snapshot, but it doesn't tell me when a shard snapshot goes from paused to starting.

**Comment by DaveCTurner** [comment]
> The pausing part would be O(#shards) messages, but how about the finalization override of shard generation? Are concurrent shard snapshots that common?

Yes, common enough in high-shard-count clusters that I wouldn't want to log this verbosely by default.

> This appears to set the current state for a shard snapshot, but it doesn't tell me when a shard snapshot goes from paused to starting again.

Right, but when the shard reports that it has paused then we call this from `org.elasticsearch.snapshots.SnapshotsService.SnapshotShardsUpdateContext.EntryContext#startShardSnapshot` which will unpause it again if the node is no longer marked for shutdown.

There's also this line if the shard is paused and then migrates to a different node:

https://github.com/elastic/elasticsearch/blob/cd8b1f9dc96565e3a781131631514f324e44109d/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L1211-L1212

**Comment by DiannaHohensee** [comment]
Okay, followed the code around [the log message you linked](https://github.com/elastic/elasticsearch/blob/cd8b1f9dc96565e3a781131631514f324e44109d/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L1211-L1212), and that makes sense to me as a clear move from PAUSED*->INIT. I upped the log level to debug, and added more details.

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java (hunk 1 above)
Comment (DiannaHohensee) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java]: Nvm, repositoryShardId.shardId() is the ShardID. I was confusing things with shard generation, again.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**